### ISSUE TITLE
Issue with Incorrect Inference Results in INT8 Model After Converting Custom YOLOv11.pt to TensorRT

### ISSUE THREAD

**Issue by xiaoche-24** [ultralytics/ultralytics#19353]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

After converting a custom-trained YOLOv11.pt model to TensorRT in FP16, the inference results are correct. However, when converting the same model to TensorRT in INT8, the inference results are incorrect, with most of the scores being zero. When using the official YOLOv11s.pt model, both INT8 and FP16 TensorRT conversions produce correct detection results. The environment and calibration dataset for all conversions are the same. What could be causing this issue?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @xiaoche-24, thank you for bringing this up and for your interest in Ultralytics 🚀! We recommend checking out the [Ultralytics Docs](https://docs.ultralytics.com/) for helpful guides and examples covering [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage that might provide insights into your issue.

To better assist you, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) of the issue. This should include relevant code snippets, scripts, or steps you used for converting your custom-trained YOLOv11.pt model to TensorRT in INT8. The additional details will help us debug and understand the root cause more effectively 🐛.

In the meantime:

## Upgrade

Ensure you're using the latest `ultralytics` package along with all dependencies. You can update everything in a [**Python>=3.8**](https://www.python.org/) environment with the following command:

```bash
pip install -U ultralytics
```

Verify the issue persists with the latest version.

## Environments

If you want to test in pre-configured environments, YOLO models can be run in the following up-to-date verified setups:  
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

## Community

You might also explore our Ultralytics community platforms for additional support:
- Join our [Discord](https://discord.com/invite/ultralytics) server for real-time discussions 🎧.
- Check out [Discourse](https://community.ultralytics.com/) for detailed, topic-driven conversations.
- Head to [Subreddit](https://reddit.com/r/Ultralytics) to connect with other users.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, it means all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml) tests are passing. Our CI workflows are scheduled daily and confirm the correct operation of YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu systems.

This is an automated response to help point you in the right direction ⚡. An Ultralytics engineer will review and assist you further as soon as possible. Thank you for your patience and understanding! 🚀

**Comment by Y-T-G** [comment]
You should be using your calibration dataset for your custom trained model. The default one is for COCO.

**Comment by xiaoche-24** [comment]
> You should be using your calibration dataset for your custom trained model. The default one is for COCO.
I used calibration data from a custom dataset, and I just found the issue. I didn't load the pre-trained weights during training, which resulted in a 36MB pt model, while the official pt model is 18MB. After trying to load the official pre-trained weights for training, I obtained a pt model size of 18MB, which matches the official model size. I was then able to successfully convert the model to an INT TensorRT model, and it produced correct inference results.

**Comment by xiaoche-24** [comment]
> You should be using your calibration dataset for your custom trained model. The default one is for COCO.

I used calibration data from a custom dataset, and I just found the issue. I didn't load the pre-trained weights during training, which resulted in a 36MB pt model, while the official pt model is 18MB. After trying to load the official pre-trained weights for training, I obtained a pt model size of 18MB, which matches the official model size. I was then able to successfully convert the model to an INT TensorRT model, and it produced correct inference results.

**Comment by Y-T-G** [comment]
The double filesize is a result of incomplete training.

**Comment by xiaoche-24** [comment]
> The double filesize is a result of incomplete training.

After completing the training, the saved pt model did indeed return to 18MB. However, after converting it to an int8 model, the targets were still not detected. Upon analyzing the output, I found that the score values had generally decreased. For instance, the fp16 model detects a target with a score of 0.88, but the int8 model detects the same target with a score of 0.36, which I only noticed after lowering the detection threshold. Could you help explain what might be causing this decrease in score values?

**Comment by Y-T-G** [comment]
Bad calibration dataset.

**Comment by xiaoche-24** [comment]
> Bad calibration dataset.

Thank you for your reply. How can I construct a suitable calibration dataset? How many images should be used for the calibration dataset when converting a TensorRT model?

**Comment by Y-T-G** [comment]
At least 300 images that are diverse.

**Comment by xiaoche-24** [comment]
> At least 300 images that are diverse.

I evaluated the accuracy of the pt model, as well as the int8 and fp16 models in TensorRT using a validation set. The results are as follows: the pt model achieved an accuracy of 86.58%, the fp16 model achieved 85.72%, and the int8 model achieved 75.58%. These results were obtained using a calibration dataset of 600 images. Notably, the int8 quantized model exhibits a significant drop in accuracy. Are there any alternative methods to reduce the accuracy loss in the quantized model and improve its performance?

**Comment by glenn-jocher** [comment]
For INT8 quantization accuracy improvements, ensure your calibration dataset closely matches inference scenarios and consider these steps:

1. **Verify Data Distribution**: Confirm calibration images represent real-world variability seen during inference
2. **Adjust Workspace/Batch**: Experiment with larger `workspace` (4-8GB) and `batch` values during export
3. **Try Layer-wise Calibration**: While we default to ENTROPY_CALIBRATION_2, you can explore custom calibration implementations

For detailed optimization strategies, see our [TensorRT INT8 quantization guide](https://docs.ultralytics.com/integrations/tensorrt/#exporting-tensorrt-with-int8-quantization). Some accuracy loss is expected with INT8, but FP16 often provides better speed/accuracy balance for many applications.
